pkg/reconciler/buildrun/resources/sources/bundle_core.go

```diff
@@ -0,0 +1,254 @@
+	"github.com/google/go-containerregistry/pkg/v1/tarball"
+)
+
+const shpIgnoreFilename = ".shpignore"
+
+// Bundle a local directory as-is into a container image. See remote.Option
+// for optional options to the image push to the registry, for example to
+// provide the appropriate access credentials.
+func Bundle(ref name.Reference, directory string, options ...remote.Option) (name.Digest, error) {
+	r, err := pack(directory)
+	if err != nil {
+		return name.Digest{}, err
+	}
+
+	bundleLayer, err := tarball.LayerFromReader(r)
+	if err != nil {
+		return name.Digest{}, err
+	}
+
+	image, err := mutate.AppendLayers(empty.Image, bundleLayer)
+	if err != nil {
+		return name.Digest{}, err
+	}
+
+	image, err = mutate.Time(image, time.Unix(0, 0))
+	if err != nil {
+		return name.Digest{}, err
+	}
+
+	hash, err := image.Digest()
+	if err != nil {
+		return name.Digest{}, err
+	}
+
+	if err := remote.Write(ref, image, options...); err != nil {
+		return name.Digest{}, err
+	}
+
+	return name.NewDigest(fmt.Sprintf("%s@%v",
+		ref.Name(),
+		hash.String(),
+	))
+}
+
+// Unbundle a container image layer content into a local directory. Analog to
+// the sources.Bundle function, optional remote.Option can be used to configure
+// settings for the image pull, i.e. access credentials.
+func Unbundle(ref name.Reference, targetPath string, options ...remote.Option) error {
+	desc, err := remote.Get(ref, options...)
+	if err != nil {
+		return err
+	}
+
+	image, err := desc.Image()
+	if err != nil {
+		return err
+	}
+
+	rc := mutate.Extract(image)
+	defer rc.Close()
+
+	return unpack(rc, targetPath)
+}
+
+func pack(directory string) (io.Reader, error) {
+	var split = func(path string) []string { return strings.Split(path, string(filepath.Separator)) }
+
+	var write = func(w io.Writer, path string) error {
+		file, err := os.Open(path)
+		if err != nil {
+			return err
+		}
+
+		defer file.Close()
+
+		_, err = io.Copy(w, file)
+		return err
+	}
+
+	var followSymLink = func(path string) (string, os.FileInfo, error) {
+		deref, err := os.Readlink(path)
+		if err != nil {
+			return "", nil, err
+		}
+
+		if !filepath.IsAbs(deref) {
+			deref = filepath.Join(
+				filepath.Dir(path),
+				deref,
+			)
+		}
+
+		info, err := os.Stat(deref)
+		return deref, info, err
+	}
+
+	var patterns []gitignore.Pattern
+	if file, err := os.Open(filepath.Join(directory, shpIgnoreFilename)); err == nil {
+		defer file.Close()
+
+		domain := split(directory)
+		scanner := bufio.NewScanner(file)
+		for scanner.Scan() {
+			line := strings.TrimSpace(scanner.Text())
+			if len(line) != 0 && !strings.HasPrefix(line, "#") {
+				patterns = append(patterns, gitignore.ParsePattern(line, domain))
+			}
+		}
+
+		if err := scanner.Err(); err != nil {
+			return nil, err
+		}
+	}
+
+	matcher := gitignore.NewMatcher(patterns)
+
+	var buf bytes.Buffer
+	var tw = tar.NewWriter(&buf)
+	defer tw.Close()
+
+	err := filepath.Walk(directory, func(path string, info os.FileInfo, err error) error {
```

Comment:
According to the doc at https://pkg.go.dev/path/filepath#Walk the newer `WalkDir` is more efficient.

Though that is new with 1.16, and we still are at 1.15 per https://github.com/shipwright-io/build/blob/main/.github/workflows/ci.yml 

I'm fine if we don't want to take on moving to 1.16 for the project in this PR, but if that is the path we take, how about a TODO comment here to switch to `WalkDir` when we do move to 1.16

Reply:
I think we are fine to move to 1.16 as 1.17 is out there. It was a nice exercise to change it to `WalkDir`. Although, I think the improvements will not be that good for us as we require the `d.Info()` call on almost all files anyway.